[WIP] Load rules lazily

diml opened the pull request:
Currently, jbuilder generates all the rules for the whole workspace at once. This can take a noticeable amount of time when the workspace is big.

This PR changes that by making jbuilder load the rules lazily, on a per-directory basis. `Build_system.create` now takes a callback `gen_rules:(dir:Path.t -> unit)` which is expected to generate all the rules for directory `dir`.

We don't enforce that the callback generates rules only for the given directory, as this makes writing the rules painful. For instance, when generating the library rules in `_build/default/src`, it is convenient to generate the doc rules in `_build/default/_doc/foo` as well. What happens is that when generating the rules for `_build/default/_doc/foo`, we just tell the system to load the `_build/default/src` directory first.

This PR adds the `Scheme` module which is used to coordinate all this. The name is imported from the `Scheme` module of Jenga which does something similar.

After this PR, we should make `Scheme` a monad. This will make things easier when we add support for dynamic rule generation. I haven't done it in this PR to minimise the diff, but we should do it before adding support for plugins.

- [x] write the code
- [x] write proper doc
- [x] handle deletion of stale directories
- [x] make sure it is faster using the jbuild-universe repository (waiting for #369)

rgrinberg commented:
Exciting. This is noticeable delay even when I'm building the jst set of packages locally.

diml commented:
Ok, this PR is almost ready. It will just need a bit more testing I think and a few changes to support deletion of stale directories. One good thing about this PR is that it makes one invariant about the order in which the rules are generated explicit.

I did the following on the jbuilder universe:

```shell
$ mkdir test
$ echo '(rule (with-stdout-to x (echo toto)))' > test/jbuild
$ jbuider build test/x
```

With master it takes 1.9s, with this PR it takes 0.6s. I have some ideas why it's still slow and I think this can be improved.

In general, I expect that build will be slightly faster as well, since the graph now has less edges. Before, the compilation of .ml/.mli files depended on all individual .cmi/.cmx files from the external lib directory. This PR introduces a bit of dynamism, so we can now generate one alias per external directory and extension on the fly.

rgrinberg commented on src/gen_rules.ml:
```diff
@@ -1143,10 +1161,30 @@ let gen ~contexts ?(filter_out_optional_stanzas_with_missing_deps=true)
-   String_map.of_alist_exn context_names_and_stanzas)
+  let context_names = List.map contexts ~f:(fun c -> c.Context.name) in
+  let merlin_context = List.find contexts ~f:(fun c -> c.merlin) in
+  let gen_rules _build_system ~dir =
```
Are we going to need this _build_system parameter elsewhere? It doesn't seem like it's used anywhere currently.

diml replied:
Nope. But given that `Build_system.create` takes the `gen_rules` callback as argument, it seems to me that the callback should get the build system as argument. It's not used in jbuilder, but eventually this stuff might be exposed as a library.

rgrinberg commented on src/alias.ml:
```diff
@@ -125,24 +126,45 @@ module Store = struct
-  type t = (Fq_name.t, entry) Hashtbl.t
+  type dir_status =
+    { mutable dir_aliases : entry list
+    ; mutable closed  : bool
```
What's the use of this `closed` flag? I see it being set and asserted but it doesn't seem to affect behavior. Is there an invariant this is meant to maintain?

diml replied:
There is indeed an invariant. I'll add a comment

rgrinberg commented on src/gen_rules.ml:
```diff
@@ -830,117 +854,106 @@ Add it to your jbuild file to remove this warning.
-              ; "version"
-              ; "VERSION"
-              ]
+      SC.on_load_dir sctx ~dir:path ~f:(fun () ->
```
Why do we need to guard the META generation behind on_load_dir?

diml replied:
just to make it lazy

diml commented:
I moved `Alias` as a sub-module of `Build_system`. It was starting to become spaghetti code and having `Alias` in `Build_system` makes things a lot simpler and cleaner since we can process the regular targets and aliases of a directory at the same time.

I did various other small refactoring in this feature, in particular the way fallback rules are
implemented is cleaner. I also changed the implementation of `--force` to no longer delete stamp files, since we can no longer delete them all in advance.

I removed the generation of `META.pkg.from-jbuilder`, nobody was using this feature and it was complicated to keep the behaviour "only generate META.pkg if there is no rule t o generate it already". Such behaviour is dodgy anyway.

In order to get a more robust system, I constrained things more:
- it is no longer possible to generate targets outside of the build contexts and `_build/install`
- it is no longer possible to generate aliases outside of the build contexts

As a result, I changed a bit the naming of the alias stamp files: instead of `_build/.aliases/_build/default/runtest-....`, it is now `_build/.aliases/default/runtest`. We win one `_build` in the filename.

The interpretation of targets on the command line is now a bit simpler as well:
- `<file>` always mean `_build/<context>/<file>` for every context where it is defined
- `@<alias>` always mean `(alias_rec _build/<context>/<alias>)` for every context where it is defined

Other improvements:
- in `--dev` mode, the rules for converting installed `.cma` files into `.cma.js` files are now longer generated eagerly for all installed packages, instead they are generated on demand
- before this feature, the compilation of each OCaml source files used to depend on all the individual .cmi/.cmx files of required libraries. Now we generate time stamp files on the fly to make the graph smaller. These files are named `_build/.misc/<hash-of-dir><extension>`
